### PR TITLE
feat: SolidStart Adapter

### PR DESCRIPTION
**Pull request by OrJDev**
Closes #

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

this adds an adapter for SolidStart,
 it is very similar to the fetchRequestHandler, yet has a different type of context, needs to be exported multiple times (`GET`, `POST`) and doesn't require the `req` and `endpoint` args - but instead gets them by its own (using the `event` property provided by solid-start).
it used to be in https://github.com/orjdev/solid-start-trpc but it is pretty messy having to install bunch of different packages so i think it should be an adapter in here aswell.

Using the fetchRequestHandler:
```ts
import { APIEvent } from "solid-start/api";
import { fetchRequestHandler } from '@trpc/server/adapters/fetch';
import { appRouter } from "~/lib/router";
const handler = (event: APIEvent) => 
  fetchRequestHandler({
    endpoint: '/api/trpc',
    req: event.request,
    router: appRouter,
    createContext: () => ({}),
  });
export const GET = handler;
export const POST = handler;
```

Using the SolidStart handler:

```ts
import { type inferAsyncReturnType, initTRPC } from '@trpc/server';
import {
  type CreateSolidContextOptions,
  createSolidApiHandler,
} from '@trpc/server/adapters/solid';

export const createContext = async (opts: CreateSolidContextOptions) => {
  return {
    req: opts.event.request,
    resHeaders: opts.resHeaders,
  };
};

export type IContext = inferAsyncReturnType<typeof createContext>;

const t = initTRPC.context<IContext>().create();
const appRouter = t.router({
  // [...]
});

export const { GET, POST } = createSolidApiHandler({
  router: appRouter,
  createContext,
});
```

### CODE REVIEW

**Review by juliusmarminge — COMMENTED** [review]
Thank you for this. Looks interesting.

I think we should have some tests for this - there are tests for all the other adapters which you can probably use as a reference point.

Would also be nice with an example I think?

**Comment by OrJDev** [comment]
> Thank for this 🙏🏼 . Looks interesting.
> 
> 
> 
> I think we should have some tests for this - there are tests for all the other adapters which you can probably use as a reference point.
> 
> 
> 
> Would also be nice with an example I think?

An example would be [create jd app](https://github.com/orjdev/create-jd-app)? 

And sure i could try to write some tests later

**Comment by juliusmarminge** [comment]
> An example would be [create jd app](https://github.com/orjdev/create-jd-app)?

Is there no more "minimal" approach just showcasing the adapter? I've never used Solid myself (although I probably should test it out soon...)

**Comment by OrJDev** [comment]
> > An example would be [create jd app](https://github.com/orjdev/create-jd-app)?
> 
> 
> 
> Is there no more "minimal" approach just showcasing the adapter? I've never used Solid myself (although I probably should test it out soon...)

I edited the message but a minimal approach is also provided within the docs, in this pr i also wrote docs for the adapter.

**Comment by OrJDev** [comment]
> > > An example would be [create jd app](https://github.com/orjdev/create-jd-app)?
> > 
> > 
> > 
> > Is there no more "minimal" approach just showcasing the adapter? I've never used Solid myself (although I probably should test it out soon...)
> 
> I edited the message but a minimal approach is also provided within the docs, in this pr i also wrote docs for the adapter. 

And yes you should definitely try it if you want to make your life easier 😎

**Comment by OrJDev** [comment]
> > > An example would be [create jd app](https://github.com/orjdev/create-jd-app)?
> > 
> > 
> > 
> > Is there no more "minimal" approach just showcasing the adapter? I've never used Solid myself (although I probably should test it out soon...)
> 
> I edited the message but a minimal approach is also provided within the docs, in this pr i also wrote docs for the adapter. 

I will just make a codesandbox after this is merged and published (no need to patch it) and make another pr to update the docs with a fully functioning example of client and server

**Comment by juliusmarminge** [comment]
@OrJDev I added the adapter to the rollup config so that it's properly exposed to the world :)

Also started an example based on the [next-minimal-starter](https://github.com/trpc/trpc/tree/main/examples/next-minimal-starter) although there's something wrong with it if you wanna take a look at it

**Comment by OrJDev** [comment]
> @OrJDev I added the adapter to the rollup config so that it's properly exposed to the world :)
> 
> 
> 
> Also started an example based on the [next-minimal-starter](https://github.com/trpc/trpc/tree/main/examples/next-minimal-starter) although there's something wrong with it if you wanna take a look at it

Yes you used ssr with the csr package, if you want to use ssr you should use solid-trpc@ssr. Just add { ssr: false } when calling the solid vite plugin in vite.config.ts

**Comment by juliusmarminge** [comment]
> Yes you used ssr with the csr package, if you want to use ssr you should use solid-trpc@ssr. Or just add { ssr: false } when calling the solid vite plugin in vite.config.ts

Aah, now I just get a white screen... Guess you can't do loading states like I did in Solid?

CleanShot 2023-01-21 at 13 36 09

🤣

**Comment by juliusmarminge** [comment]
> > > Yes you used ssr with the csr package, if you want to use ssr you should use solid-trpc@ssr. Or just add { ssr: false } when calling the solid vite plugin in vite.config.ts
> 
> > 
> 
> > Aah, now I just get a white screen... Guess you can't do loading states like I did in Solid?
> 
> > 
> 
> > CleanShot 2023-01-21 at 13 36 09
> 
> > 
> 
> > Too used with React 🤣 Feel free to fix it ;)
> 
> 
> 
> client side rendering is not recommended at all in solid-start, but when using tanstack query you should be using switches - i deleted your example and created a new working example with ssr.  feel free to modify the markdown / whatever you want (;

Yea we'll make it more minimal imo, just the basics to get it running like the next-minimal-starter.

Can do it later if you haven't done it already

**Comment by OrJDev** [comment]
Screenshot 2023-01-21 at 15 00 20

looks pretty minimal to me, what do you think

File: examples/solidstart-minimal-starter/package.json
Comment (juliusmarminge) [inline examples/solidstart-minimal-starter/package.json:15]: Why not tanstack?

Comment (OrJDev) [reply examples/solidstart-minimal-starter/package.json:15]: it is a patched package that supports ssr published by the creator of tanstack solid query (aryan),
he published it so i could add ssr support to trpc,
that is because ssr will be supported under the tanstack org only in version 5 (which may take a while)

Comment (OrJDev) [reply examples/solidstart-minimal-starter/package.json:15]: https://github.com/orjdev/solid-trpc

https://github.com/orjdev/solid-trpc/tree/ssr

therefore it is labeled under the `@ssr` tag for now

**Comment by juliusmarminge** [comment]
Looks good to me now apart from missing tests for the adapter:

CleanShot 2023-01-21 at 15 05 57

**Comment by OrJDev** [comment]
Is the test required or could it be merged (looking at the example provided)

**Comment by juliusmarminge** [comment]
> Is the test required or could it be merged (looking at the example provided)

Hold off for now, we're debating on whether we want to maintain more adapters in core. Sorry bout that...

**Comment by OrJDev** [comment]
> > Is the test required or could it be merged (looking at the example provided)
> 
> 
> 
> Hold off for now, we're debating on whether we want to maintain more adapters in core. Sorry bout that...

No problem its not urgent as it is published under solid-start-trpc for now

**Review by KATT — CHANGES_REQUESTED** [review]
We don't include third party plugins here, can you instead have an empty folder with a markdown document leading to your official one? Makes it easier to maintain only 1 source of truth

Example: https://github.com/trpc/trpc/tree/main/examples/trpc-chrome

**Comment by OrJDev** [comment]
done

**Comment by thejskhan** [comment]
Any updates for this? I've just started to setup SolidStart with tRPC and Prisma. I'm eager to hear any developments moving forward

**Comment by sachinraja** [comment]
@thejskhan you should be able to use [the Fetch adapter](https://trpc.io/docs/fetch)

**Comment by OrJDev** [comment]
> Any updates for this? I've just started to setup SolidStart with tRPC and Prisma. I'm eager to hear any developments moving forward

use [create jd app](https://github.com/orjdev/create-jd-app) choose trpc and prisma (;
solid-start-trpc already exists this pr was created to make this an official adapter, it shouldn't be an issue

**Comment by OrJDev** [comment]
> Hey, thanks for this, @OrJDev!
> 
> 
> 
> - This is missing tests
> 
> - Is it possible to use the fetch adapter instead as @sachinraja asked?
> 
> 

As my pr description says, while the fetch adapter works solid has a different kind of context (api event), this also figures out stuff by its own and creates an actual handler, more mentioned in the description.

Regarding the test, i could make one but first lets see that you guys are interested in that.

**Comment by OrJDev** [comment]
> > Hey, thanks for this, @OrJDev!
> > 
> > 
> > 
> > - This is missing tests
> > 
> > - Is it possible to use the fetch adapter instead as @sachinraja asked?
> > 
> > 
> 
> As my pr description says, while the fetch adapter works solid has a different kind of context (api event), this also figures out stuff by its own and creates an actual handler, more mentioned in the description.
> 
> Regarding the test, i could make one but first lets see that you guys are interested in that. 

Like this makes it much cleaner and provides much better dx

```ts
export const { GET, POST } = createSolidApiHandler({
  router: appRouter,
  createContext,
});
```

**Comment by juliusmarminge** [comment]
How stable is the SolidStart way of doing this? Will it change if it starts usign Astro under the hood? Might be worth holding off and just recommending the fetch adapter in the meantime? Perhaps a docs addition similar to #3767?

**Comment by OrJDev** [comment]
> How stable is the SolidStart way of doing this? Will it change if it starts usign Astro under the hood? Might be worth holding off and just recommending the fetch adapter in the meantime? Perhaps a docs addition similar to #3767?

Both astro and solidstart are just web request & response, even if it changes to astro we don't need to change any code on our end, maybe just lowercase the "GET / POST" export.

**Comment by juliusmarminge** [comment]
Idk using the fetch adapter seems pretty clean?

https://github.com/juliusmarminge/trpc-solid/blob/main/src/routes/api/trpc/%5Btrpc%5D.ts

**Comment by OrJDev** [comment]
> Using the fetch adapter seems pretty clean?
> 
> 
> 
> https://github.com/juliusmarminge/trpc-solid/blob/main/src/routes/api/trpc/%5Btrpc%5D.ts
> 
> 
> 
> 

Im not saying it is not clean, it is pretty clean but mine is cleaner 😎, there is always a better solution

**Comment by juliusmarminge** [comment]
Added docs for how to use it with the fetch adapter: https://trpc.io/docs/fetch#solidstart

Given that, we don't think it's worth maintaining a separate adapter for Solid when the integration story is the same as Astro, Remix etc

**Pull request closed without merging**
